<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Detect tissue heterogeneity in expression profiles with gene sets • BioQC</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Detect tissue heterogeneity in expression profiles with gene sets">
<meta property="og:description" content="BioQC performs quality control of high-throughput expression data
    based on tissue gene signatures. It can detect tissue heterogeneity in gene
    expression data. The core algorithm is a Wilcoxon-Mann-Whitney test that is
    optimised for high performance.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">BioQC</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.19.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="articles/BioQC.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/bioqc-efficiency.html">BioQC Algorithm: Speeding up the Wilcoxon-Mann-Whitney Test</a>
    </li>
    <li>
      <a href="articles/bioqc-introduction.html">BioQC: Detect tissue heterogeneity in gene expression data</a>
    </li>
    <li>
      <a href="articles/bioqc-signedGenesets.html">Using BioQC with signed genesets</a>
    </li>
    <li>
      <a href="articles/bioqc-simulation.html">BioQC-benchmark: Testing Efficiency, Sensitivity and Specificity of BioQC on simulated and real-world data</a>
    </li>
    <li>
      <a href="articles/bioqc-wmw-test-performance.html">Comparing the Wilcoxon-Mann-Whitney to alternative statistical tests</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">


<p>BioQC is a is a R/<a href="https://bioconductor.org/packages/release/bioc/html/BioQC.html">Bioconductor</a> package to detect tissue heterogeneity in gene expression data. Tissue heterogeneity is a consequence of unintended profiling of cells of other origins than the tissue of interest and can have both technical (<em>e.g.</em> imperfect disection) or biological (<em>e.g.</em> immune infiltration) reasons.</p>
<p>We <a href="https://www.biorxiv.org/content/10.1101/2020.12.02.407809v1">demonstrated</a> that tissue heterogeneity is prevalent in 5-15% of all gene expression studies. Ignoring tissue heterogeneity reduces statistical power of data analysis and can, in the worst case, invalidate the conclusions of a study. Therefore, we propose applying BioQC as a routine step in every gene-expression analysis pipeline.</p>
<p>The BioQC method is described in</p>
<blockquote>
<p>Zhang, Jitao David, Klas Hatje, Gregor Sturm, Clemens Broger, Martin Ebeling, Martine Burtin, Fabiola Terzi, Silvia Ines Pomposiello, and Laura Badi. “Detect Tissue Heterogeneity in Gene Expression Data with BioQC.” BMC Genomics 18 (2017): 277. <a href="https://doi.org/10.1186/s12864-017-3661-2">doi:10.1186/s12864-017-3661-2</a>).</p>
</blockquote>
<div id="basic-usage" class="section level2">
<h2 class="hasAnchor">
<a href="#basic-usage" class="anchor"></a>Basic Usage</h2>
<p>BioQC implements a computationally efficient Wilcoxon-Mann-Whitney test for testing for enrichment of tissue signatures. A database of 150 tissue signatures derived from large-scale transcriptomics studies is shipped with the BioQC package.</p>
<p>To apply BioQC to a <code>genes x samples</code> gene expression matrix, run:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://accio.github.io/BioQC">BioQC</a></span><span class="op">)</span>

<span class="co"># load the tissue signatures</span>
<span class="va">gmtFile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/exp.tissuemark.affy.roche.symbols.gmt"</span>,
                        package<span class="op">=</span><span class="st">"BioQC"</span><span class="op">)</span>
<span class="va">gmt</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/readGmt.html">readGmt</a></span><span class="op">(</span><span class="va">gmtFile</span><span class="op">)</span>

<span class="co"># perform BioQC enrichment test on a gene expression matrix</span>
<span class="va">bioqc_res</span> <span class="op">=</span> <span class="fu"><a href="reference/wmwTest.html">wmwTest</a></span><span class="op">(</span><span class="va">expr_mat</span>, <span class="va">gmt</span><span class="op">)</span>
<span class="va">bioqc_scores</span> <span class="op">=</span> <span class="fu"><a href="reference/absLog10p.html">absLog10p</a></span><span class="op">(</span><span class="va">bioqc_res</span><span class="op">)</span></code></pre></div>
<p>The following figure shows the BioQC scores from the <a href="https://accio.github.io/BioQC/bioqc.html">kidney example</a> visualized as heatmap. We note that in samples 23-25 <em>adipose</em> and <em>pancreas</em> signatures have been detected, hinting at a containation with those tissues. For this dataset, we could validate the contamination with qPCR.</p>
<p><img src="reference/figures/kidney_heatmap.png" alt="example heatmap"></p>
<p>For a more detailed example explaining how to use other data structures or custom signatures see</p>
<ul>
<li>
<a href="https://accio.github.io/BioQC/bioqc.html">Applying BioQC to a real-world kidney dataset</a>.</li>
<li>
<a href="https://accio.github.io/BioQC/articles/bioqc-introduction.html">Introduction to BioQC</a>.</li>
</ul>
<p>For advanced usages, check out:</p>
<ul>
<li><a href="https://accio.github.io/BioQC/articles/bioqc-signedGenesets.html">Single sample gene set enrichment analysis with BioQC</a></li>
</ul>
<p>For benchmarks and details about the algorithm, see:</p>
<ul>
<li><a href="https://accio.github.io/BioQC/bioqc-simulation.html">BioQC-benchmark: Testing Efficiency, Sensitivity and Specificity of BioQC on simulated and real-world data</a></li>
<li><a href="https://accio.github.io/BioQC/articles/bioqc-wmw-test-performance.html">Comparing the Wilcoxon-Mann-Whitney to alternative statistical tests</a></li>
<li><a href="https://accio.github.io/BioQC/articles/bioqc-efficiency.html">The BioQC algorithm: speeding up the WMW-test</a></li>
</ul>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<div id="bioconductor" class="section level3">
<h3 class="hasAnchor">
<a href="#bioconductor" class="anchor"></a>Bioconductor</h3>
<p>BioQC is available <a href="https://www.bioconductor.org/packages/release/bioc/html/BioQC.html">from Bioconductor</a>.</p>
<p>You can install it using</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>

<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"BioQC"</span><span class="op">)</span></code></pre></div>
</div>
<div id="bioconda" class="section level3">
<h3 class="hasAnchor">
<a href="#bioconda" class="anchor"></a>Bioconda</h3>
<p>Alternatively, you can use the <a href="https://docs.conda.io/en/latest/miniconda.html">conda</a> package manager.</p>
<ol>
<li><p>Make sure you <a href="https://bioconda.github.io/user/install.html#set-up-channels">set-up the Bioconda channel correctly</a>. The order of the channels is important!</p></li>
<li>
<p>(Optional) Create and activate an environment for BioQC</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1"></a><span class="ex">conda</span> create -n bioqc</span>
<span id="cb3-2"><a href="#cb3-2"></a><span class="ex">conda</span> activate bioqc</span></code></pre></div>
</li>
<li>
<p>Install the <code>bioconductor-bioqc</code> package in your current environment</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1"></a><span class="ex">conda</span> install bioconductor-bioqc</span></code></pre></div>
</li>
</ol>
</div>
<div id="from-github" class="section level3">
<h3 class="hasAnchor">
<a href="#from-github" class="anchor"></a>From Github</h3>
<p>The easiest way to install the development version from GitHub is using the <code>remotes</code> package:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span>
<span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span><span class="op">(</span><span class="st">"accio/BioQC"</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="contact" class="section level2">
<h2 class="hasAnchor">
<a href="#contact" class="anchor"></a>Contact</h2>
<p>If you have questions regarding BioQC or want to report a bug, please use the <a href="https://github.com/accio/BioQC/issues">issue tracker</a>.</p>
<p>Alternatively you can reach out to Jitao David Zhang via <a href="jitao_david.zhang@roche.com">e-mail</a>.</p>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Download from BIOC at <br><a href="https://www.bioconductor.org/packages/BioQC">https://​www.bioconductor.org/​packages/​BioQC</a>
</li>
<li>Report a bug at <br><a href="https://accio.github.io/BioQC/issues">https://​accio.github.io/​BioQC/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>GPL (&gt;=3) + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>
<div class="citation">
<h2>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html">Citing BioQC</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Jitao David Zhang <br><small class="roles"> Maintainer, author </small>  </li>
<li>Laura Badi <br><small class="roles"> Author </small>  </li>
<li>Gregor Sturm <br><small class="roles"> Author </small>  </li>
<li>Roland Ambs <br><small class="roles"> Author </small>  </li>
<li>Iakov Davydov <br><small class="roles"> Author </small>  </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/Accio/BioQC/actions"><img src="https://github.com/Accio/BioQC/workflows/check-bioc/badge.svg" alt="R build status"></a></li>
<li><a href="https://codecov.io/gh/Accio/BioQC"><img src="https://codecov.io/gh/Accio/BioQC/branch/master/graph/badge.svg" alt="codecov"></a></li>
<li><a href="http://bioconda.github.io/recipes/bioconductor-bioqc/README.html"><img src="https://img.shields.io/badge/install%20with-bioconda-brightgreen.svg?style=flat" alt="install with bioconda"></a></li>
<li><a href="https://github.com/accio/BioQC/blob/master/LICENSE.md"><img src="https://img.shields.io/badge/license-GPLv3-green.svg" alt="license"></a></li>
<li><a href="https://accio.github.io/BioQC"><img src="https://img.shields.io/badge/docs-pkgdown-blue.svg" alt="docs"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Jitao David Zhang, Laura Badi, Gregor Sturm, Roland Ambs, Iakov Davydov.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
