<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using BioQC with signed genesets • BioQC</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Using BioQC with signed genesets">
<meta property="og:description" content="BioQC">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">BioQC</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.19.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/BioQC.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/bioqc-efficiency.html">BioQC Algorithm: Speeding up the Wilcoxon-Mann-Whitney Test</a>
    </li>
    <li>
      <a href="../articles/bioqc-introduction.html">BioQC: Detect tissue heterogeneity in gene expression data</a>
    </li>
    <li>
      <a href="../articles/bioqc-signedGenesets.html">Using BioQC with signed genesets</a>
    </li>
    <li>
      <a href="../articles/bioqc-simulation.html">BioQC-benchmark: Testing Efficiency, Sensitivity and Specificity of BioQC on simulated and real-world data</a>
    </li>
    <li>
      <a href="../articles/bioqc-wmw-test-performance.html">Comparing the Wilcoxon-Mann-Whitney to alternative statistical tests</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="bioqc-signedGenesets_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Using BioQC with signed genesets</h1>
                        <h4 class="author">Jitao David Zhang and Gregor Sturm</h4>
            
            <h4 class="date">2020-12-06</h4>
      
      
      <div class="hidden name"><code>bioqc-signedGenesets.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>Besides being used as a QC-tool for gene expression data, <em>BioQC</em> can be used for general-purpose gene-set enrichment analysis. Its core function, <em>wmwTest</em>, can handle not only “normal”, unsigned gene sets, but also signed genesets where two sets of genes represent signatures that are positively and negatively regulated respectively.</p>
<p>This vignette describes an example of such applications. First we load the BioQC library.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://accio.github.io/BioQC">BioQC</a></span><span class="op">)</span></code></pre></div>
</div>
<div id="gmt-read-in" class="section level2">
<h2 class="hasAnchor">
<a href="#gmt-read-in" class="anchor"></a>GMT read-in</h2>
<p>We first open a toy GMT file containing signed genesets with <em>readSignedGmt</em>. The function reads the file into a <em>SignedGenesets</em> object.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gmtFile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/test.gmt"</span>, package<span class="op">=</span><span class="st">"BioQC"</span><span class="op">)</span>
<span class="co">## print the file context</span>
<span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/readLines.html">readLines</a></span><span class="op">(</span><span class="va">gmtFile</span><span class="op">)</span>, sep<span class="op">=</span><span class="st">"\n"</span><span class="op">)</span></code></pre></div>
<pre><code>## GS_A TestGeneSet_A   AKT1    AKT2    AKT3
## GS_B TestGeneSet_B   MAPK1   MAPK3   MAPK8
## GS_C_UP  TestGeneSet_C   ERBB2   ERBB3
## GS_C_DN  TestGeneSet_C   EGFR    ERBB4
## GS_D_UP  TestGeneSet_D   GATA2   GATA4
## GS_D_DN  TestGeneSet_D   GATA1   GATA3
## GS_E_DN  TestGeneSet_E   TSC1    TSC2</code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## read the file into SignedGenesets</span>
<span class="va">genesets</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readSignedGmt.html">readSignedGmt</a></span><span class="op">(</span><span class="va">gmtFile</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">genesets</span><span class="op">)</span></code></pre></div>
<pre><code>## A list of 5 signed gene-sets
##   GS_A (no genes)
##   GS_B (no genes)
##   GS_C[n=4]
##     positive[n=2]:ERBB2,ERBB3
##     negative[n=2]:EGFR,ERBB4
##   GS_D[n=4]
##     positive[n=2]:GATA2,GATA4
##     negative[n=2]:GATA1,GATA3
##   GS_E[n=2]
##     positive[n=0]:
##     negative[n=2]:TSC1,TSC2</code></pre>
<p>Note that though the GMT file contains 6 non-empty lines, only five signed genesets are returned by <em>readSignedGmt</em> because a pair of negative and negative geneset of GS_C is available.</p>
<p>Note that in its current form, no genes are reported for GS_A and GS_B, because the names of both sets do not match the patterns. User can decide how to handle such genesets that match patterns of neither positive nor negative sets (GS_A and GS_B in this case). By default, such genesets are ignored; however, user can decide to treat them as either positive or negative genesets. For the purpose of this tutorial, we treat these genesets as positive.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">genesets</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readSignedGmt.html">readSignedGmt</a></span><span class="op">(</span><span class="va">gmtFile</span>, nomatch<span class="op">=</span><span class="st">"pos"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">genesets</span><span class="op">)</span></code></pre></div>
<pre><code>## A list of 5 signed gene-sets
##   GS_A[n=3]
##     positive[n=3]:AKT1,AKT2,AKT3
##     negative[n=0]:
##   GS_B[n=3]
##     positive[n=3]:MAPK1,MAPK3,MAPK8
##     negative[n=0]:
##   GS_C[n=4]
##     positive[n=2]:ERBB2,ERBB3
##     negative[n=2]:EGFR,ERBB4
##   GS_D[n=4]
##     positive[n=2]:GATA2,GATA4
##     negative[n=2]:GATA1,GATA3
##   GS_E[n=2]
##     positive[n=0]:
##     negative[n=2]:TSC1,TSC2</code></pre>
</div>
<div id="expression-object-construction" class="section level2">
<h2 class="hasAnchor">
<a href="#expression-object-construction" class="anchor"></a>Expression object construction</h2>
<p>Next we construct an ExpressionSet object containing 100 genes, with some of the genes in the GMT file present.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">1887</span><span class="op">)</span>
<span class="va">testN</span> <span class="op">&lt;-</span> <span class="fl">100L</span>
<span class="va">testSampleCount</span> <span class="op">&lt;-</span> <span class="fl">3L</span>
<span class="va">testGenes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"AKT1"</span>, <span class="st">"AKT2"</span>, <span class="st">"ERBB2"</span>, <span class="st">"ERBB3"</span>, <span class="st">"EGFR"</span>,<span class="st">"TSC1"</span>, <span class="st">"TSC2"</span>, <span class="st">"GATA2"</span>, <span class="st">"GATA4"</span>, <span class="st">"GATA1"</span>, <span class="st">"GATA3"</span><span class="op">)</span>
<span class="va">testRows</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">testGenes</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Gene"</span>, <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">testGenes</span><span class="op">)</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span><span class="op">:</span><span class="va">testN</span>, sep<span class="op">=</span><span class="st">""</span><span class="op">)</span><span class="op">)</span>
<span class="va">testMatrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="va">testN</span><span class="op">*</span><span class="va">testSampleCount</span>, sd<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span>, nrow<span class="op">=</span><span class="va">testN</span>, dimnames<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">testRows</span>, <span class="cn">NULL</span><span class="op">)</span><span class="op">)</span>
<span class="va">testMatrix</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>,<span class="op">]</span> <span class="op">&lt;-</span> <span class="va">testMatrix</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>,<span class="op">]</span><span class="op">+</span><span class="fl">10</span>
<span class="va">testMatrix</span><span class="op">[</span><span class="fl">6</span><span class="op">:</span><span class="fl">7</span>,<span class="op">]</span> <span class="op">&lt;-</span> <span class="va">testMatrix</span><span class="op">[</span><span class="fl">6</span><span class="op">:</span><span class="fl">7</span>,<span class="op">]</span><span class="op">-</span><span class="fl">10</span>
<span class="va">testMatrix</span><span class="op">[</span><span class="fl">3</span><span class="op">:</span><span class="fl">4</span>,<span class="op">]</span> <span class="op">&lt;-</span> <span class="va">testMatrix</span><span class="op">[</span><span class="fl">3</span><span class="op">:</span><span class="fl">4</span>,<span class="op">]</span><span class="op">-</span><span class="fl">5</span>
<span class="va">testMatrix</span><span class="op">[</span><span class="fl">5</span>,<span class="op">]</span> <span class="op">&lt;-</span> <span class="va">testMatrix</span><span class="op">[</span><span class="fl">5</span>,<span class="op">]</span><span class="op">+</span><span class="fl">5</span>
<span class="va">testEset</span> <span class="op">&lt;-</span> <span class="fu">new</span><span class="op">(</span><span class="st">"ExpressionSet"</span>, exprs<span class="op">=</span><span class="va">testMatrix</span><span class="op">)</span></code></pre></div>
<p>The expression of somes genes are delibrately changed so that some genesets are expected to be significantly enriched.</p>
<ul>
<li>AKT1 and AKT2 are higher expressed than the background genes (‘background’ hereafter).</li>
<li>TSC1 and TSC2 are lower expressed than the background.</li>
<li>ERBB2 and ERBB3 are moderately lower expressed than the background.</li>
<li>EGFR is moderately higher expressed than the background.</li>
</ul>
<p>Considering the geneset composition printed above, we can expect that * GS_A shall be significantly up-regulated; * GS_B shall not be significant, because its genes are missing; * GS_C shall be significantly down-regulated, with positive targets down-regulated and negative targets up-regulated; * GS_D shall not always be significant, because its genes are comparable to the background; * GS_E shall be significantly up-regulated, because its negative targets are down-regulated.</p>
</div>
<div id="match-genes" class="section level2">
<h2 class="hasAnchor">
<a href="#match-genes" class="anchor"></a>Match genes</h2>
<p>With both <em>SignedGenesets</em> and <em>ExpressionSet</em> objects at hand, we can query the indices of genes of genesets in the <em>ExpressionSet</em> object.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">testIndex</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/matchGenes.html">matchGenes</a></span><span class="op">(</span><span class="va">genesets</span>, <span class="va">testEset</span>, col<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">testIndex</span><span class="op">)</span></code></pre></div>
<pre><code>## A list of 5 signed indices with offset=1
## Options: NA removed: TRUE; duplicates removed: TRUE
##   GS_A[n=2]
##     positive[n=2]:1,2
##     negative[n=0]:
##   GS_B (no genes)
##   GS_C[n=3]
##     positive[n=2]:3,4
##     negative[n=1]:5
##   GS_D[n=4]
##     positive[n=2]:8,9
##     negative[n=2]:10,11
##   GS_E[n=2]
##     positive[n=0]:
##     negative[n=2]:6,7</code></pre>
<p>If the <em>ExpressionSet</em> object has GeneSymbols in its featureData other than the feature names, user can specify the parameter <em>col</em> in <em>matchGenes</em> to make the match working. <em>matchGene</em> also works with characters and matrix as input, please check out the help pages for such uses.</p>
</div>
<div id="perform-the-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#perform-the-analysis" class="anchor"></a>Perform the analysis</h2>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">wmwResult.greater</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/wmwTest.html">wmwTest</a></span><span class="op">(</span><span class="va">testEset</span>, <span class="va">testIndex</span>, valType<span class="op">=</span><span class="st">"p.greater"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">wmwResult.greater</span><span class="op">)</span></code></pre></div>
<pre><code>##                1           2           3
## GS_A 0.008185751 0.008185751 0.008185751
## GS_B 1.000000000 1.000000000 1.000000000
## GS_C 0.997942914 0.997942914 0.997942914
## GS_D 0.924153922 0.114118870 0.211756278
## GS_E 0.008185751 0.008185751 0.008185751</code></pre>
<p>As expected, GS_A and GS_E are significant when the alternative hypothesis is <em>greater</em>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">wmwResult.less</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/wmwTest.html">wmwTest</a></span><span class="op">(</span><span class="va">testEset</span>, <span class="va">testIndex</span>, valType<span class="op">=</span><span class="st">"p.less"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">wmwResult.less</span><span class="op">)</span></code></pre></div>
<pre><code>##                1           2           3
## GS_A 0.992348913 0.992348913 0.992348913
## GS_B 1.000000000 1.000000000 1.000000000
## GS_C 0.002192387 0.002192387 0.002192387
## GS_D 0.078389208 0.889240837 0.793302033
## GS_E 0.992348913 0.992348913 0.992348913</code></pre>
<p>As expected, GS_C is significant when the alternative hypothesis is <em>less</em>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">wmwResult.two.sided</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/wmwTest.html">wmwTest</a></span><span class="op">(</span><span class="va">testEset</span>, <span class="va">testIndex</span>, valType<span class="op">=</span><span class="st">"p.two.sided"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">wmwResult.two.sided</span><span class="op">)</span></code></pre></div>
<pre><code>##                1           2           3
## GS_A 0.016371502 0.016371502 0.016371502
## GS_B 1.000000000 1.000000000 1.000000000
## GS_C 0.004384774 0.004384774 0.004384774
## GS_D 0.156778415 0.228237741 0.423512555
## GS_E 0.016371502 0.016371502 0.016371502</code></pre>
<p>As expected, GS_A, GS_C, and GS_E are significant when the alternative hypothesis is <em>two.sided</em>.</p>
<p>A simple way to integrate the results of both alternative hypotheses into one numerical value is the <em>Q</em> value, which is defined as absolute log10 transformation of the smaller p value of the two hypotheses, times the sign defined by <em>greater</em>=1 and <em>less</em>=-1.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">wmwResult.Q</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/wmwTest.html">wmwTest</a></span><span class="op">(</span><span class="va">testEset</span>, <span class="va">testIndex</span>, valType<span class="op">=</span><span class="st">"Q"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">wmwResult.Q</span><span class="op">)</span></code></pre></div>
<pre><code>##               1          2          3
## GS_A  1.7859115  1.7859115  1.7859115
## GS_B  0.0000000  0.0000000  0.0000000
## GS_C -2.3580528 -2.3580528 -2.3580528
## GS_D -0.8047137  0.6416125  0.3731337
## GS_E  1.7859115  1.7859115  1.7859115</code></pre>
<p>As expected, genesets GS_A, GS_C, and GS_E have large absolute values, suggesting they are potentially enriched; while GS_A and GS_E are positively enriched, GS_C is negatively enriched.</p>
</div>
<div id="conclusions" class="section level2">
<h2 class="hasAnchor">
<a href="#conclusions" class="anchor"></a>Conclusions</h2>
<p>In summary, <em>BioQC</em> can be used to perform gene-set enrichment analysis with signed genesets. Its speed is comparable to that of unsigned genesets.</p>
</div>
<div id="acknowledgement" class="section level2">
<h2 class="hasAnchor">
<a href="#acknowledgement" class="anchor"></a>Acknowledgement</h2>
<p>I would like to thank the authors of the gCMAP package with the idea of integrating signed genesets into the framework of Wilcoxon-Mann-Whitney test.</p>
</div>
<div id="r-session-info" class="section level2">
<h2 class="hasAnchor">
<a href="#r-session-info" class="anchor"></a>R session info</h2>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## R Under development (unstable) (2020-12-04 r79554)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.1 LTS
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats     graphics  grDevices utils     datasets  methods  
## [8] base     
## 
## other attached packages:
## [1] BioQC_1.19.2        Biobase_2.51.0      BiocGenerics_0.37.0
## [4] knitr_1.30         
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_1.0.5         magrittr_2.0.1     edgeR_3.33.0       lattice_0.20-41   
##  [5] R6_2.5.0           ragg_0.4.0         rlang_0.4.9        stringr_1.4.0     
##  [9] tools_4.1.0        grid_4.1.0         xfun_0.19          htmltools_0.5.0   
## [13] systemfonts_0.3.2  yaml_2.2.1         assertthat_0.2.1   rprojroot_2.0.2   
## [17] digest_0.6.27      pkgdown_1.6.1.9000 crayon_1.3.4       textshaping_0.2.1 
## [21] fs_1.5.0           memoise_1.1.0      evaluate_0.14      rmarkdown_2.5     
## [25] limma_3.47.1       stringi_1.5.3      compiler_4.1.0     desc_1.2.0        
## [29] locfit_1.5-9.4</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Jitao David Zhang, Laura Badi, Gregor Sturm, Roland Ambs, Iakov Davydov.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
