<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>BioQC-kidney: The kidney expression example • BioQC</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="BioQC-kidney: The kidney expression example">
<meta property="og:description" content="BioQC">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">BioQC</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.19.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/BioQC.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/bioqc-efficiency.html">BioQC Algorithm: Speeding up the Wilcoxon-Mann-Whitney Test</a>
    </li>
    <li>
      <a href="../articles/bioqc-introduction.html">BioQC: Detect tissue heterogeneity in gene expression data</a>
    </li>
    <li>
      <a href="../articles/bioqc-signedGenesets.html">Using BioQC with signed genesets</a>
    </li>
    <li>
      <a href="../articles/bioqc-simulation.html">BioQC-benchmark: Testing Efficiency, Sensitivity and Specificity of BioQC on simulated and real-world data</a>
    </li>
    <li>
      <a href="../articles/bioqc-wmw-test-performance.html">Comparing the Wilcoxon-Mann-Whitney to alternative statistical tests</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="BioQC_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>BioQC-kidney: The kidney expression example</h1>
            
            <h4 class="date">2020-12-06</h4>
      
      
      <div class="hidden name"><code>BioQC.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>In this vignette, we demonstrate the use of <em>BioQC</em> with a case study where mouse kidney samples were profiled for gene expression. Results of BioQC pointed to potential tissue heterogeneity caused by pancreas contamination which was confirmed by qRT-PCR experiments.</p>
</div>
<div id="importing-the-data" class="section level2">
<h2 class="hasAnchor">
<a href="#importing-the-data" class="anchor"></a>Importing the data</h2>
<p>Let’s first load the BioQC signatures using <code>readGmt</code>:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://accio.github.io/BioQC">BioQC</a></span><span class="op">)</span>
<span class="va">gmtFile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/exp.tissuemark.affy.roche.symbols.gmt"</span>,
                        package<span class="op">=</span><span class="st">"BioQC"</span><span class="op">)</span>
<span class="va">gmt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readGmt.html">readGmt</a></span><span class="op">(</span><span class="va">gmtFile</span><span class="op">)</span></code></pre></div>
<p>The example data for this vignette is shipped as part of the BioQC package. Alternatively, it is available in the <a href="https://github.com/accio/BioQC/inst/extdata/kidney_example.rda">GitHub repository</a>. The expression data is stored in a <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/Biobase/inst/doc/ExpressionSetIntroduction.pdf">ExpressionSet</a>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/kidney_example.rda"</span>, package<span class="op">=</span><span class="st">"BioQC"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="va">file</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">eset</span><span class="op">)</span></code></pre></div>
<p>The dataset contains expression of 34719 genes in 25 samples. The expression profile was normalized with RMA normalization. The signals were also <code>log2</code>-transformed; however, this step does not affect the result of <em>BioQC</em> since it is essentially a non-parametric statistical test.</p>
</div>
<div id="running-bioqc" class="section level2">
<h2 class="hasAnchor">
<a href="#running-bioqc" class="anchor"></a>Running BioQC</h2>
<p>Next we run the core function of the <em>BioQC</em> package, <code>wmwTest</code>, to perform the analysis.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span><span class="op">(</span><span class="va">bioqcRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/wmwTest.html">wmwTest</a></span><span class="op">(</span><span class="va">eset</span>, <span class="va">gmt</span>,
                                 valType<span class="op">=</span><span class="st">"p.greater"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>##    user  system elapsed 
##   0.855   0.059   0.916</code></pre>
<p>The function returns <em>one-sided</em> <span class="math inline">\(p\)</span>-values of Wilcoxon-Mann-Whitney test.</p>
<p>For better visualization we * filter the p-value matrix (a sensible cutoff depends on the actual dataset and the expected signatures); * transform the p-values using <span class="math inline">\(|\log_{10} p|\)</span>. We refer to thse transformed values as <em>BioQC score</em>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bioqcResFil</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filterPmat.html">filterPmat</a></span><span class="op">(</span><span class="va">bioqcRes</span>, <span class="fl">1E-6</span><span class="op">)</span>
<span class="va">bioqcAbsLogRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/absLog10p.html">absLog10p</a></span><span class="op">(</span><span class="va">bioqcResFil</span><span class="op">)</span></code></pre></div>
<p>We inspect the BioQC scores by visualizing them as Heatmap. By closer examination we find that the expression of pancreas and adipose specific genes is significantly enriched in samples 23-25:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">RColorBrewer</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/stats/heatmap.html">heatmap</a></span><span class="op">(</span><span class="va">bioqcAbsLogRes</span>, Colv<span class="op">=</span><span class="cn">NA</span>, Rowv<span class="op">=</span><span class="cn">TRUE</span>,
        cexRow<span class="op">=</span><span class="fl">0.85</span>, scale<span class="op">=</span><span class="st">"row"</span>,
        col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rev.html">rev</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">brewer.pal</a></span><span class="op">(</span><span class="fl">7</span>, <span class="st">"RdBu"</span><span class="op">)</span><span class="op">)</span>,
        labCol<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">bioqcAbsLogRes</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="BioQC_files/figure-html/heatmap-1.png" alt="BioQC scores (defined as abs(log10(p))) of the samples visualized in heatmap. Red and blue indicate high and low scores respectively." width="768" style="display:block; margin: auto"><p class="caption">
BioQC scores (defined as abs(log10(p))) of the samples visualized in heatmap. Red and blue indicate high and low scores respectively.
</p>
</div>
<p>Visual inspection reveals that there might be contaminations in samples 23-25, potentially by pancreas and adipose tissue.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">filRes</span> <span class="op">&lt;-</span> <span class="va">bioqcAbsLogRes</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Kidney_NGS_RNASEQATLAS_0.6_3"</span>,
                           <span class="st">"Pancreas_Islets_NR_0.7_3"</span><span class="op">)</span>,<span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/matplot.html">matplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">filRes</span><span class="op">)</span>, pch<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"K"</span>, <span class="st">"P"</span><span class="op">)</span>, type<span class="op">=</span><span class="st">"b"</span>, lty<span class="op">=</span><span class="fl">1L</span>,
        ylab<span class="op">=</span><span class="st">"BioQC score"</span>, xlab<span class="op">=</span><span class="st">"Sample index"</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="BioQC_files/figure-html/vis_bioqc-1.png" alt="BioQC scores (defined as abs(log10(p))) of the samples. K and P represent kidney and pancreas signature scores respectively." width="768" style="display:block; margin: auto"><p class="caption">
BioQC scores (defined as abs(log10(p))) of the samples. K and P represent kidney and pancreas signature scores respectively.
</p>
</div>
</div>
<div id="validation-with-quantitative-rt-pcr" class="section level2">
<h2 class="hasAnchor">
<a href="#validation-with-quantitative-rt-pcr" class="anchor"></a>Validation with quantitative RT-PCR</h2>
<p>To confirm the hypothesis generated by <em>BioQC</em>, we performed qRT-PCR experiments to test two pancreas-specific genes’ expression in the same set of samples. Note that the two genes (amylase and elastase) are not included in the signature set provided by BioQC.</p>
<p>The results are shown in the figure below. It seems likely that sample 23-25 are contaminated by nearby pancreas tissues when the kidney was dissected. Potential contamination by adipose tissues remains to be tested.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">amylase</span> <span class="op">&lt;-</span> <span class="va">eset</span><span class="op">$</span><span class="va">Amylase</span>
<span class="va">elastase</span> <span class="op">&lt;-</span> <span class="va">eset</span><span class="op">$</span><span class="va">Elastase</span>
<span class="va">pancreasScore</span> <span class="op">&lt;-</span> <span class="va">bioqcAbsLogRes</span><span class="op">[</span><span class="st">"Pancreas_Islets_NR_0.7_3"</span>,<span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>, mar<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3</span>,<span class="fl">3</span>,<span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>, mgp<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">1</span>,<span class="fl">0</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">amylase</span><span class="op">~</span><span class="va">pancreasScore</span>, log<span class="op">=</span><span class="st">"y"</span>, pch<span class="op">=</span><span class="fl">21</span>, bg<span class="op">=</span><span class="st">"red"</span>,
     xlab<span class="op">=</span><span class="st">"BioQC pancreas score"</span>, ylab<span class="op">=</span><span class="st">"Amylase"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/text.html">text</a></span><span class="op">(</span><span class="va">pancreasScore</span><span class="op">[</span><span class="fl">23</span><span class="op">:</span><span class="fl">25</span><span class="op">]</span>,<span class="va">amylase</span><span class="op">[</span><span class="fl">23</span><span class="op">:</span><span class="fl">25</span><span class="op">]</span>, <span class="fl">23</span><span class="op">:</span><span class="fl">25</span>, pos<span class="op">=</span><span class="fl">1</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">elastase</span><span class="op">~</span><span class="va">pancreasScore</span>, log<span class="op">=</span><span class="st">"y"</span>, pch<span class="op">=</span><span class="fl">21</span>, bg<span class="op">=</span><span class="st">"red"</span>,
     xlab<span class="op">=</span><span class="st">"BioQC pancreas score"</span>, ylab<span class="op">=</span><span class="st">"Elastase"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/text.html">text</a></span><span class="op">(</span><span class="va">pancreasScore</span><span class="op">[</span><span class="fl">23</span><span class="op">:</span><span class="fl">25</span><span class="op">]</span>,<span class="va">elastase</span><span class="op">[</span><span class="fl">23</span><span class="op">:</span><span class="fl">25</span><span class="op">]</span>, <span class="fl">23</span><span class="op">:</span><span class="fl">25</span>, pos<span class="op">=</span><span class="fl">1</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="BioQC_files/figure-html/rt_pcr_result-1.png" alt="Correlation between qRT-PCR results and BioQC pancreas score" width="768" style="display:block; margin: auto"><p class="caption">
Correlation between qRT-PCR results and BioQC pancreas score
</p>
</div>
</div>
<div id="impact-of-sample-removal-on-differential-gene-expression-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#impact-of-sample-removal-on-differential-gene-expression-analysis" class="anchor"></a>Impact of sample removal on differential gene expression analysis</h2>
<p>In this study, four mice of the <em>FVB/NJ</em> strain received nephrectomy operation and treatment of Losartan, an angiotensin II receptor antagonist drug, and four mice reveiced an sham operation and Losartan. Within the Nephrectomy+Losartan group, one sample (index 24) is possibly contaminated by pancreas. Suppose now we are interested in the differential gene expression between the conditions. We now run the analysis twice, once with and once without the contaminated sample, to study the impact of removing heterogenous samples detected by <em>BioQC</em>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioinf.wehi.edu.au/limma">limma</a></span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">isNeph</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html">with</a></span><span class="op">(</span><span class="fu">pData</span><span class="op">(</span><span class="va">eset</span><span class="op">)</span>, <span class="va">Strain</span><span class="op">==</span><span class="st">"FVB/NJ"</span> <span class="op">&amp;</span>
                 <span class="va">TREATMENTNAME</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Nephrectomy-Losartan"</span>, <span class="st">"Sham-Losartan"</span><span class="op">)</span><span class="op">)</span>
<span class="va">isContam</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html">with</a></span><span class="op">(</span><span class="fu">pData</span><span class="op">(</span><span class="va">eset</span><span class="op">)</span>, <span class="va">INDIVIDUALNAME</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"BN7"</span>, <span class="st">"FNL8"</span>, <span class="st">"FN6"</span><span class="op">)</span><span class="op">)</span>
<span class="va">esetNephContam</span> <span class="op">&lt;-</span> <span class="va">eset</span><span class="op">[</span>,<span class="va">isNeph</span><span class="op">]</span>
<span class="va">esetNephExclContam</span> <span class="op">&lt;-</span> <span class="va">eset</span><span class="op">[</span>, <span class="va">isNeph</span> <span class="op">&amp;</span> <span class="op">!</span><span class="va">isContam</span><span class="op">]</span>
<span class="va">getDEG</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">eset</span><span class="op">)</span> <span class="op">{</span>
  <span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">eset</span><span class="op">$</span><span class="va">TREATMENTNAME</span>, levels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Sham-Losartan"</span>,<span class="st">"Nephrectomy-Losartan"</span><span class="op">)</span><span class="op">)</span>
  <span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="va">group</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">design</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ShamLo"</span>, <span class="st">"NephLo"</span><span class="op">)</span>
  <span class="va">contrast</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/makeContrasts.html">makeContrasts</a></span><span class="op">(</span>contrasts<span class="op">=</span><span class="st">"NephLo"</span>, levels<span class="op">=</span><span class="va">design</span><span class="op">)</span>
  <span class="fu">exprs</span><span class="op">(</span><span class="va">eset</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/normalizebetweenarrays.html">normalizeBetweenArrays</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="fu">exprs</span><span class="op">(</span><span class="va">eset</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
  <span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/lmFit.html">lmFit</a></span><span class="op">(</span><span class="va">eset</span>, design<span class="op">=</span><span class="va">design</span><span class="op">)</span>
  <span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/contrasts.fit.html">contrasts.fit</a></span><span class="op">(</span><span class="va">fit</span>, <span class="va">contrast</span><span class="op">)</span>
  <span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html">eBayes</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span>
  <span class="va">tt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/toptable.html">topTable</a></span><span class="op">(</span><span class="va">fit</span>, n<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">eset</span><span class="op">)</span><span class="op">)</span>
  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">tt</span><span class="op">)</span>
<span class="op">}</span>

<span class="va">esetNephContam.topTable</span> <span class="op">&lt;-</span> <span class="fu">getDEG</span><span class="op">(</span><span class="va">esetNephContam</span><span class="op">)</span>
<span class="va">esetNephExclContam.topTable</span> <span class="op">&lt;-</span> <span class="fu">getDEG</span><span class="op">(</span><span class="va">esetNephExclContam</span><span class="op">)</span>
<span class="va">esetFeats</span> <span class="op">&lt;-</span> <span class="fu">featureNames</span><span class="op">(</span><span class="va">eset</span><span class="op">)</span>
<span class="va">esetNephTbl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>feature<span class="op">=</span><span class="va">esetFeats</span>,
                          OrigGeneSymbol<span class="op">=</span><span class="va">esetNephContam.topTable</span><span class="op">[</span><span class="va">esetFeats</span>,<span class="op">]</span><span class="op">$</span><span class="va">OrigGeneSymbol</span>,
                          GeneSymbol<span class="op">=</span><span class="va">esetNephContam.topTable</span><span class="op">[</span><span class="va">esetFeats</span>,<span class="op">]</span><span class="op">$</span><span class="va">GeneSymbol</span>,
                          Contam.logFC<span class="op">=</span><span class="va">esetNephContam.topTable</span><span class="op">[</span><span class="va">esetFeats</span>,<span class="op">]</span><span class="op">$</span><span class="va">logFC</span>,
                          ExclContam.logFC<span class="op">=</span><span class="va">esetNephExclContam.topTable</span><span class="op">[</span><span class="va">esetFeats</span>,<span class="op">]</span><span class="op">$</span><span class="va">logFC</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>, mar<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3</span>,<span class="fl">3</span>,<span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span><span class="op">+</span><span class="fl">0.5</span>, mgp<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">1</span>,<span class="fl">0</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/with.html">with</a></span><span class="op">(</span><span class="va">esetNephTbl</span>, <span class="fu"><a href="https://rdrr.io/r/graphics/smoothScatter.html">smoothScatter</a></span><span class="op">(</span><span class="va">Contam.logFC</span><span class="op">~</span><span class="va">ExclContam.logFC</span>,
                                xlab<span class="op">=</span><span class="st">"Excluding one contaminating sample [logFC]"</span>,
                                ylab<span class="op">=</span><span class="st">"Including one contaminating sample [logFC]"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>
<span class="va">isDiff</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html">with</a></span><span class="op">(</span><span class="va">esetNephTbl</span>, <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">Contam.logFC</span><span class="op">-</span><span class="va">ExclContam.logFC</span><span class="op">)</span><span class="op">&gt;=</span><span class="fl">2</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/with.html">with</a></span><span class="op">(</span><span class="va">esetNephTbl</span>, <span class="fu"><a href="https://rdrr.io/r/graphics/points.html">points</a></span><span class="op">(</span><span class="va">Contam.logFC</span><span class="op">[</span><span class="va">isDiff</span><span class="op">]</span><span class="op">~</span><span class="va">ExclContam.logFC</span><span class="op">[</span><span class="va">isDiff</span><span class="op">]</span>, pch<span class="op">=</span><span class="fl">16</span>, col<span class="op">=</span><span class="st">"red"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="BioQC_files/figure-html/deg-1.png" alt="Log2 fold change (logFC) values reported by *limma* with one contaminated sample included (y-axis) or excluded (x-axis). Genes strongly affected by the contamination are indicated by red dots." width="480" style="display:block; margin: auto"><p class="caption">
Log2 fold change (logFC) values reported by <em>limma</em> with one contaminated sample included (y-axis) or excluded (x-axis). Genes strongly affected by the contamination are indicated by red dots.
</p>
</div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">diffTable</span> <span class="op">&lt;-</span> <span class="va">esetNephTbl</span><span class="op">[</span><span class="va">isDiff</span>,<span class="op">]</span>
<span class="va">diffGenes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">diffTable</span><span class="op">[</span>,<span class="st">"GeneSymbol"</span><span class="op">]</span><span class="op">)</span>
<span class="va">pancreasSignature</span> <span class="op">&lt;-</span> <span class="va">gmt</span><span class="op">[[</span><span class="st">"Pancreas_Islets_NR_0.7_3"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">genes</span>
<span class="va">diffGenesPancreas</span> <span class="op">&lt;-</span> <span class="va">diffGenes</span> <span class="op">%in%</span> <span class="va">pancreasSignature</span>
<span class="va">diffTable</span><span class="op">$</span><span class="va">isPancreasSignature</span> <span class="op">&lt;-</span> <span class="va">diffTable</span><span class="op">$</span><span class="va">GeneSymbol</span> <span class="op">%in%</span> <span class="va">pancreasSignature</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">diffTable</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Probeset"</span>, <span class="st">"GeneSymbol"</span>, <span class="st">"Human ortholog"</span>,
                         <span class="st">"Log2FC"</span>, <span class="st">"Log2FC (excl. contam.)"</span>,
                         <span class="st">"IsPancreasSignature"</span><span class="op">)</span>
<span class="va">diffTable</span> <span class="op">&lt;-</span> <span class="va">diffTable</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">diffTable</span><span class="op">$</span><span class="va">Log2FC</span>, decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>,<span class="op">]</span></code></pre></div>
<p>We found that 22 probesets representing 17 genes are associated with much stronger expression changes if the contaminated sample is not excluded (table below). Not surprisingly almost all of these genes are highly expressed in normal human pancreas tissues, and 13 genes belong to the pancreas signature used by <em>BioQC</em>.</p>
<p>In summary, we observe that tissue heterogeneity can impact down-stream analysis results and negatively affect reproducibility of gene expression data if it remains overlooked. It underlines again the value of applying <em>BioQC</em> as a first-line quality control tool.</p>
</div>
<div id="r-session-info" class="section level2">
<h2 class="hasAnchor">
<a href="#r-session-info" class="anchor"></a>R Session Info</h2>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## R Under development (unstable) (2020-12-04 r79554)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.1 LTS
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats     graphics  grDevices utils     datasets  methods  
## [8] base     
## 
## other attached packages:
## [1] limma_3.47.1        RColorBrewer_1.1-2  BioQC_1.19.2       
## [4] Biobase_2.51.0      BiocGenerics_0.37.0 knitr_1.30         
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_1.0.5         magrittr_2.0.1     edgeR_3.33.0       lattice_0.20-41   
##  [5] R6_2.5.0           ragg_0.4.0         rlang_0.4.9        highr_0.8         
##  [9] stringr_1.4.0      tools_4.1.0        grid_4.1.0         xfun_0.19         
## [13] KernSmooth_2.23-18 htmltools_0.5.0    systemfonts_0.3.2  yaml_2.2.1        
## [17] assertthat_0.2.1   rprojroot_2.0.2    digest_0.6.27      pkgdown_1.6.1.9000
## [21] crayon_1.3.4       textshaping_0.2.1  fs_1.5.0           memoise_1.1.0     
## [25] evaluate_0.14      rmarkdown_2.5      stringi_1.5.3      compiler_4.1.0    
## [29] desc_1.2.0         locfit_1.5-9.4</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Jitao David Zhang, Laura Badi, Gregor Sturm, Roland Ambs, Iakov Davydov.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
