<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Comparing the Wilcoxon-Mann-Whitney to alternative statistical tests • BioQC</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Comparing the Wilcoxon-Mann-Whitney to alternative statistical tests">
<meta property="og:description" content="BioQC">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">BioQC</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.21.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/BioQC.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/bioqc-efficiency.html">BioQC Algorithm: Speeding up the Wilcoxon-Mann-Whitney Test</a>
    </li>
    <li>
      <a href="../articles/bioqc-introduction.html">BioQC: Detect tissue heterogeneity in gene expression data</a>
    </li>
    <li>
      <a href="../articles/bioqc-signedGenesets.html">Using BioQC with signed genesets</a>
    </li>
    <li>
      <a href="../articles/bioqc-simulation.html">BioQC-benchmark: Testing Efficiency, Sensitivity and Specificity of BioQC on simulated and real-world data</a>
    </li>
    <li>
      <a href="../articles/bioqc-wmw-test-performance.html">Comparing the Wilcoxon-Mann-Whitney to alternative statistical tests</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="bioqc-wmw-test-performance_files/header-attrs-2.9/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Comparing the Wilcoxon-Mann-Whitney to alternative statistical tests</h1>
            
            <h4 data-toc-skip class="date">2021-07-19</h4>
      
      
      <div class="hidden name"><code>bioqc-wmw-test-performance.Rmd</code></div>

    </div>

    
    
<p>In this document, we show that the Wilcoxon-Mann-Whitney test is comparable or superior to alternative methods.</p>
<p>Two alternative methods could be compared with the Wilcoxon-Mann-Whitney (WMW) test proposed by BioQC: the Kolmogorov-Smirnov (KS) test, and the Student’s t-test, or more particularly, the Welch’s test which does not assume equal sample number or equal variance, which is appropriate in the setting of gene expression studies.</p>
<ol style="list-style-type: decimal">
<li>It is documented in statistics literature that the WMW test offers a higher power than the Kolmogorov-Smirnov test<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>,<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a>.</li>
<li>Compared with parameterized test methods such as the t-test, the WMW test is (a) resistance to monotone transformation, (b) suffers less from outliers, and (c) provides higher efficiency when many genes are profiled and the distribution of gene expression deviates from the normal distribution, which are important criteria in genome-wide expression data.</li>
</ol>
<p>Based on these considerations, BioQC implements a computationally efficient version of the WMW test. In order not to confuse end-users, no alternative methods are implemented.</p>
<p>Nevertheless, in order to demonstrate the power of WMW test in comparison with the KS-test or the t-test, we performed the sensitivity benchmark described in the <a href="bioqc-simulation.html">simulation studies</a>, for the two alternative tests respectively.</p>
<div class="figure" style="text-align: center">
<img src="bioqc-wmw-test-performance_files/figure-html/sensitivity_benchmark_fig-1.png" alt="**Figure 1:** Sensitivity benchmark. Expression levels of genes in the ovary signature are dedicately sampled randomly from normal distributions with different mean values. The lines show the enrichment score for the Wilcoxon-Mann-Whitney test, the t-test and the Kolmogorov-Smirnov test respectively. In the right panel, outliers were added by adding a random value to 1% of the simulated genes. " width="768" style="display:block; margin: auto"><p class="caption">
<strong>Figure 1:</strong> Sensitivity benchmark. Expression levels of genes in the ovary signature are dedicately sampled randomly from normal distributions with different mean values. The lines show the enrichment score for the Wilcoxon-Mann-Whitney test, the t-test and the Kolmogorov-Smirnov test respectively. In the right panel, outliers were added by adding a random value to 1% of the simulated genes.
</p>
</div>
<p>As expected, the results suggest, that both the KS-test and the WMW-test are robust to noise, while the performance of the t-test drops significantly on noisy data. Additionally, the WMW-test appears to be superior to the KS-test for low expression differences.</p>
<div id="computational-performance" class="section level2">
<h2 class="hasAnchor">
<a href="#computational-performance" class="anchor" aria-hidden="true"></a>Computational Performance</h2>
<p>Since the KS-test is so slow, we did not replicate the sensitivity benchmark from the <a href="bioqc-simulation.html">simulation studies</a> using the enrichment score rank. While it takes BioQC about 4 seconds on a single thread to test all 155 signatures, it already takes the KS-test about 2 seconds to test a single signature.</p>
<pre><code>##       test replications elapsed relative
## 2  runKS()            5  10.027    1.000
## 1 runWMW()            5  17.511    1.746</code></pre>
</div>
<div id="r-session-info" class="section level2">
<h2 class="hasAnchor">
<a href="#r-session-info" class="anchor" aria-hidden="true"></a>R Session Info</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## R version 4.1.0 (2021-05-18)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.2 LTS
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats4    stats     graphics  grDevices utils     datasets  methods  
## [8] base     
## 
## other attached packages:
##  [1] rbenchmark_1.0.0      ggplot2_3.3.5         plyr_1.8.6           
##  [4] reshape2_1.4.4        gplots_3.1.1          gridExtra_2.3        
##  [7] latticeExtra_0.6-29   lattice_0.20-44       hgu133plus2.db_3.13.0
## [10] org.Hs.eg.db_3.13.0   AnnotationDbi_1.55.1  IRanges_2.27.0       
## [13] S4Vectors_0.31.0      BioQC_1.21.0          Biobase_2.53.0       
## [16] BiocGenerics_0.39.1   testthat_3.0.4        knitr_1.33           
## 
## loaded via a namespace (and not attached):
##  [1] httr_1.4.2             sass_0.4.0             edgeR_3.35.0          
##  [4] bit64_4.0.5            jsonlite_1.7.2         gtools_3.9.2          
##  [7] bslib_0.2.5.1          highr_0.9              blob_1.2.1            
## [10] GenomeInfoDbData_1.2.6 yaml_2.2.1             pillar_1.6.1          
## [13] RSQLite_2.2.7          glue_1.4.2             limma_3.49.1          
## [16] digest_0.6.27          RColorBrewer_1.1-2     XVector_0.33.0        
## [19] colorspace_2.0-2       htmltools_0.5.1.1      pkgconfig_2.0.3       
## [22] zlibbioc_1.39.0        scales_1.1.1           jpeg_0.1-8.1          
## [25] tibble_3.1.2           KEGGREST_1.33.0        farver_2.1.0          
## [28] ellipsis_0.3.2         withr_2.4.2            cachem_1.0.5          
## [31] magrittr_2.0.1         crayon_1.4.1           memoise_2.0.0         
## [34] evaluate_0.14          fansi_0.5.0            fs_1.5.0              
## [37] textshaping_0.3.5      tools_4.1.0            lifecycle_1.0.0       
## [40] stringr_1.4.0          munsell_0.5.0          locfit_1.5-9.4        
## [43] Biostrings_2.61.1      compiler_4.1.0         pkgdown_1.6.1.9001    
## [46] jquerylib_0.1.4        GenomeInfoDb_1.29.3    caTools_1.18.2        
## [49] systemfonts_1.0.2      rlang_0.4.11           grid_4.1.0            
## [52] RCurl_1.98-1.3         labeling_0.4.2         bitops_1.0-7          
## [55] rmarkdown_2.9          gtable_0.3.0           DBI_1.1.1             
## [58] R6_2.5.0               utf8_1.2.1             fastmap_1.1.0         
## [61] bit_4.0.4              rprojroot_2.0.2        ragg_1.1.3            
## [64] KernSmooth_2.23-20     desc_1.3.0             stringi_1.7.2         
## [67] Rcpp_1.0.7             vctrs_0.3.8            png_0.1-7             
## [70] xfun_0.24</code></pre>
</div>
<div id="references" class="section level2">
<h2 class="hasAnchor">
<a href="#references" class="anchor" aria-hidden="true"></a>References</h2>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>Irizarry, Rafael A., et al. “Gene set enrichment analysis made simple.”Statistical methods in medical research 18.6 (2009): 565-575.<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>Filion, Guillaume J. “The signed Kolmogorov-Smirnov test: why it should not be used.” GigaScience 4.1 (2015): 1.<a href="#fnref2" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jitao David Zhang, Laura Badi, Gregor Sturm, Roland Ambs, Iakov Davydov.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link external-link">pkgdown</a> 1.6.1.9001.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
